sycl/doc/design/PropagateCompilerFlagsToRuntime.md

```diff
@@ -0,0 +1,124 @@
+was not able to set a breakpoint inside device code.
+
+## Requirements
+
+In order to support module-level debuggability, the user will compile different
+module files with different levels of optimization. These optimization levels
+must be preserved and made use of during the backend compilation. Following are
+the requirements for this feature.
+- If the user specifies `-Ox` as a front-end compile option for a particular
+module, this option must be preserved during backend JIT compilation.
+- If the user specifies `-O0` option, SYCL runtime needs to pass the appropriate
+backend option to JIT compilation stages.
+
+The following table specifies the appropriate backend options for level-zero and
+OpenCL backends.
+
+| Front-end option | L0 backend option | OpenCL backend option |
+| ---------------- | ----------------- | --------------------- |
+|      -O0         |  -ze-opt-disable  |   -cl-opt-disable     |
+|      -O1         |  -ze-opt-level=1  |   /* no option */     |
+|      -O2         |  -ze-opt-level=1  |   /* no option */     |
+|      -O3         |  -ze-opt-level=2  |   /* no option */     |
+
+
+## Proposed design
+
+This chapter discusses changes required in various stages of the compilation
+pipeline.
+
+
+### Changes to the clang front-end
+
+For each SYCL kernel, we add a new function attribute that is named
+`sycl-optlevel`. Value of this attribute is set to the optimization level which
+was used to compile the overlying module.
+
+### Changes to the sycl-post-link tool
+
+During device code split performed in the `sycl-post-link` tool, optimization
+level attribute `sycl-optlevel` is treated as an optional feature,
+i.e. device code split algorithm ensures that no kernels with different values
+of sycl-optlevel are bundled into the same device image. See also optional
+kernel features [design document](https://github.com/intel/llvm/blob/sycl/sycl/doc/design/OptionalDeviceFeatures.md#changes-to-the-post-link-tool).
+The `sycl-post-link` tool also adds a new property into the 
+`SYCL/misc properties` property set for each device code module. This entry will
+be used to store the optimization level. Name of this property is 'optLevel' and
+the value is stored as a 32-bit integer. If there is a module where user did not
+specify an optimization module, there is no new entry in the property set.
+
+### Changes to the SYCL runtime
+
+In the SYCL runtime, the device image properties can be accessed to extract the
+associated optimization level. Once the optimization level is available, a query
+is made to the plugin to identify the correct backend option. This backend
+option is added to the existing list of compiler options and is sent to the
+backend.
+
+### Changes to the plugin
+
+A new plugin API has been added. It takes the optimization level as input in
+integer format and returns `pi_result`. The signature is as follows:
+
+`pi_result` piPluginGetBackendOptimizationOption(int OptLevel);
```

Comment:
Doesn't this need to return a string somehow, which is the "backend option" noted above?

Reply:
Good catch. I fixed this in latest commit. Thanks